src/instance/mod.rs

```diff
@@ -58,13 +68,178 @@ impl Instance {
     /// on WebAssembly bytes (represented by the Python bytes type).
     #[new]
     #[allow(clippy::new_ret_no_self)]
-    fn new(object: &PyRawObject, bytes: &PyAny) -> PyResult<()> {
+    fn new(
+        object: &PyRawObject,
+        bytes: &PyAny,
+        imported_functions: &'static PyDict,
+    ) -> PyResult<()> {
+        let py = object.py();
+
         // Read the bytes.
         let bytes = <PyBytes as PyTryFrom>::try_from(bytes)?.as_bytes();
 
+        let mut import_object = ImportObject::new();
+        let mut host_function_references = Vec::with_capacity(imported_functions.len());
+
+        for (namespace_name, namespace) in imported_functions.iter() {
+            let namespace_name = namespace_name
+                .downcast_ref::<PyString>()
+                .map_err(|_| RuntimeError::py_err("Namespace name must be a string.".to_string()))?
+                .to_string()?;
+
+            let mut import_namespace = Namespace::new();
+
+            for (function_name, function) in namespace
+                .downcast_ref::<PyDict>()
+                .map_err(|_| RuntimeError::py_err("Namespace must be a dictionnary.".to_string()))?
+                .into_iter()
+            {
+                let function_name = function_name
+                    .downcast_ref::<PyString>()
+                    .map_err(|_| {
+                        RuntimeError::py_err("Function name must be a string.".to_string())
+                    })?
+                    .to_string()?;
+
+                if !function.is_callable() {
+                    return Err(RuntimeError::py_err(format!(
+                        "Function for `{}` is not callable.",
+                        function_name
+                    )));
+                }
+
+                if !function.hasattr("__annotations__")? {
+                    return Err(RuntimeError::py_err(format!(
+                        "Function `{}` must have type annotations for parameters and results.",
+                        function_name
+                    )));
+                }
+
+                let mut input_types = vec![];
+                let mut output_types = vec![];
+
+                for (name, value) in function
+                    .getattr("__annotations__")?
```

Comment:
Rather than getting the annotations from the imports, I think we should just inspect the module imports and cast to what is required. Failing in case the function signature is not expected to what we need (argument values).

Reply:
`master` now contains all the tooling to achieve that.